drud/ddev
# Move database to docker volume instead of in ~/.ddev, fixes #714

## The Problem/Issue/Bug:

OP #714 - We've had no end of trouble bind-mounting the database onto the host on various platforms, especially Windows. 

## How this PR Solves The Problem:

* Switch to using docker volume
* Provide database snapshot technique, including snapshots on ddev rm -R

TODO:
- [x] Command to create a snapshot
- [x] Command to restore a snapshot
- [x] Option on --remove-data to create snapshot (or omit it)
- [x] Figure out how to migrate existing databases into a volume.
- [x] Test scripts should delete all ddev volumes before starting.
- [x] Too much text output when running mariabackup (creating snapshot)
- [x] mariadb container tests need to use volume, not bind mount.
- [x] Update minimum docker version to 18.06
- [x] Docs required
     - Update explanation of where db is stored
     - Update remove docs

## Manual Testing Instructions:

* **Before testing, please `docker pull drud/ddev-dbserver:20180801_volume`**
* **Before testing: Any docker-compose.*.yaml files need to be changed to `version: '3.6'`, since we were forced to do a version bump here. This is an important thing to point out in release notes and announcements.
* Sample d7 databases are provided at [d7tester_test1](https://github.com/rfay/ddev/raw/4b2a65150c29a6194779ac002a8ca32565766bca/pkg/ddevapp/testdata/revert_snapshot/d7tester_test_1.sql.gz) (one node, site name "d7tester 1") and [d7tester_test2](https://github.com/rfay/ddev/raw/4b2a65150c29a6194779ac002a8ca32565766bca/pkg/ddevapp/testdata/revert_snapshot/d7tester_test_2.sql.gz) (two nodes, site name "d7tester 2")
* Create a project, load a db
* Verify that the docker volume has been created with `docker volume ls`
* Use ddev remove --remove-data to remove the project, you should see a snapshot created and the volume destroyed.
* Test `ddev snapshot`, verify that the snapshot gets created.
* Load a different site (like d7tester_test2)
* Test `ddev revert-to-snapshot` and you should lose the change you made.
* Run `ddev rm` (with or without --remove-data) and you should see a snapshot created.
* Test migration. Create a project with ddev v1.0.0 and then do a `ddev start` with it using this version. Verify that the migration happens successfully and that you have the project you expected.

This may be sensitive to different OSs; although we have tests running on all supported OSs, it will be good to get manual testing experience with this on Windows and Linux (and WIndows 10 Home).

## Automated Testing Overview:

* TestDbMigration
* TestDdevRestoreSnapshot

And of course every database-related action tests the behavior.

Required tests:
- [x] snapshot creation
- [x] snapshot loading/reverting
- [x] Full test suite needs to be run at least manually on macOS with docker edge.

## Related Issue Link(s):

OP #714 
[Mariadb upgrade to 10.2 #947](https://github.com/drud/ddev/issues/947)

## Release/Deployment notes:

## Questions for discussion

- [ ]  Currently, `ddev rm` always creates a snapshot unless you do `ddev remove --create-snapshot=false`. Think that's the right behavior?
- [ ] It may be worth figuring out how to automatically update people's docker-compose.*.yaml to `version: '3.6'`, but I'm afraid this would be a fair bit of dead code for a few people's short-term need.
